src/compile.ts

```diff
@@ -101,42 +226,103 @@ export function compile(
       );
 
       function visitor(node: ts.Node): ts.Node | ts.Node[] {
-        const visit = () => {
-          if (checker.isAppsyncResolver(node)) {
-            return visitAppsyncResolver(node);
-          } else if (checker.isAppsyncField(node)) {
-            return visitAppsyncField(node);
-          } else if (checker.isNewStepFunction(node)) {
-            return visitStepFunction(node);
-          } else if (checker.isReflectFunction(node)) {
-            return errorBoundary(() =>
-              toFunction(NodeKind.FunctionDecl, node.arguments[0])
-            );
-          } else if (checker.isEventBusWhenFunction(node)) {
-            return visitEventBusWhen(node);
-          } else if (checker.isRuleMapFunction(node)) {
-            return visitEventBusMap(node);
-          } else if (checker.isNewRule(node)) {
-            return visitRule(node);
-          } else if (checker.isNewEventTransform(node)) {
-            return visitEventTransform(node);
-          } else if (checker.isNewFunctionlessFunction(node)) {
-            return visitFunction(node);
-          } else if (checker.isApiIntegration(node)) {
-            return visitApiIntegration(node);
-          }
-          return node;
-        };
-        // keep processing the children of the updated node.
-        return ts.visitEachChild(visit(), visitor, ctx);
+        if (
+          checker.isApiIntegration(node) ||
+          checker.isAppsyncField(node) ||
+          checker.isAppsyncResolver(node) ||
+          checker.isEventBusWhenFunction(node) ||
+          checker.isNewEventTransform(node) ||
+          checker.isNewFunctionlessFunction(node) ||
+          checker.isNewRule(node) ||
+          checker.isNewStepFunction(node) ||
+          checker.isReflectFunction(node) ||
+          checker.isRuleMapFunction(node)
+        ) {
+          // if this is a Functionless primitive, transform its arguments to wrap functions in their AST form
+          return transform(node);
+        } else if (
+          ts.isCallExpression(node) &&
+          ts.isPropertyAccessExpression(node.expression) &&
+          ts.isIdentifier(node.expression.name) &&
+          node.expression.name.text === "bind"
+        ) {
+          // potentially a func.bind(self, ..args) call
+          // wrap in the generated bind function wrapper
+          return ts.factory.createCallExpression(bindName, undefined, [
+            node.expression.expression,
+            ...node.arguments,
+          ]);
+        } else if (ts.isBlock(node)) {
```

Comment:
What if the function decl is not in a block? like statements in a swtich block?

Reply:
A FunctionDeclaration is always in a block - it's a statement. If it is in an expression-position, the it will be a FunctionExpression.

Reply:
This particular piece of code is detecting function declarations (that only exist in a block) and hoisting a register call to the top of the block.

Function and Arrow expressions are detected during an ordinary visit because they can be registered in-line.